File: eventLoop.js
const Heap = require('heap');
const Promise = require('./Promise');

class EventLoop {
    constructor() {
        this.timerHeap = new Heap((a, b) => a.time - b.time);
        this.intervals = new Map();
        this.timeouts = new Map();
        this.running = false;
        this.nextIntervalId = 1;
        this.nextTimeoutId = 1;
    }

    start() {
        this.running = true;
        this.runLoop();
    }

    stop() {
        this.running = false;
    }

    runLoop() {
        while (this.running) {
            const now = Date.now();

            // Timers Phase
            while (!this.timerHeap.empty() && this.timerHeap.peek().time <= now) {
                const timer = this.timerHeap.pop();
                timer.callback();
                // If the timer is an interval timer, reschedule it
                if (timer.intervalId) {
                    this.rescheduleInterval(timer.intervalId, timer.interval);
                }
            }
        }
    }

    setTimeout(callback, delay) {
        if (typeof callback !== 'function' || delay < 0) {
            throw new Error('Invalid callback or delay');
        }
        const time = Date.now() + delay;
        const timer = { callback, time, timeoutId: this.nextTimeoutId++ };
        this.timerHeap.push(timer);
        this.timeouts.set(timer.timeoutId, timer);
        return timer.timeoutId;
    }

    clearTimeout(timeoutId) {
        this.timeouts.delete(timeoutId);
    }

    setInterval(callback, interval) {
        if (typeof callback !== 'function' || interval <= 0) {
            throw new Error('Invalid callback or interval');
        }
        const intervalId = this.nextIntervalId++;
        const intervalTimer = { callback, time: Date.now() + interval, interval, intervalId };
        this.intervals.set(intervalId, intervalTimer);
        this.timerHeap.push(intervalTimer);
        return intervalId;
    }

    clearInterval(intervalId) {
        this.intervals.delete(intervalId);
    }

    rescheduleInterval(intervalId, interval) {
        const intervalTimer = this.intervals.get(intervalId);
        if (intervalTimer) {
            intervalTimer.time = Date.now() + interval; // Adjusted to prevent drift
            this.timerHeap.push(intervalTimer);
        }
    }
}

// Example usage:
const eventLoop = new EventLoop();

// Example 1: Using setInterval and clearInterval
const intervalId = eventLoop.setInterval(() => {
    console.log("Interval callback");
}, 1000);

// eventLoop.setTimeout(() => {
//     eventLoop.clearInterval(intervalId);
// }, 5000);

// Example 2: Using setTimeout and clearTimeout
const timeoutId = eventLoop.setTimeout(() => {
    console.log("Timeout callback");
}, 2000);

// eventLoop.setTimeout(() => {
//     eventLoop.clearTimeout(timeoutId);
// }, 1000);

// Example 3: Using a custom promise with setTimeout
function fetchDataAsync() {
    return new Promise((resolve, reject) => {
        eventLoop.setTimeout(() => {
            const data = { message: "Data fetched successfully" };
            resolve(data);
        }, 5000); // Simulate 5 seconds delay for fetching data
    });
}

function processDataAsync() {
    return new Promise((resolve, reject) => {
        eventLoop.setTimeout(() => {
            const data = { message: "Data processed successfully" };
            resolve(data);
        }, 2000); // Simulate 2 seconds delay
    });
}

// Enqueue the async task and start the event loop
fetchDataAsync()
    .then(data => {
        console.log("Success:", data);
        return processDataAsync();
    })
    .then(data => {
        console.log("Success:", data);
    })
    .catch(error => {
        console.error("Error:", error);
    });


// Start the event loop
eventLoop.start();

File: eventloop.py
import os
import socket
import select
import time
from heapq import heappush, heappop

class EventLoop:
    def __init__(self):
        self.readers = {}  # Sockets with associated callbacks for read events
        self.timer_heap = []  # Min-heap for managing timer events
        self.running = False
        self.wakeup_read, self.wakeup_write = os.pipe()  # Pipe for waking up the loop

    def start(self):
        self.running = True
        self.run_loop()

    def stop(self):
        self.running = False
        self._wakeup()

    def run_loop(self):
        while self.running:
            # Calculate the timeout for the next scheduled timer event
            timeout = self._calculate_timeout()
            
            # I/O Events Phase: Wait for I/O events or timeout using select
            readable, _, _ = select.select(list(self.readers.keys()) + [self.wakeup_read], [], [], timeout)
            
            if self.wakeup_read in readable:
                # Clear the wakeup signal if wakeup pipe is ready
                os.read(self.wakeup_read, 1)
            
            # Handle I/O events
            self._process_ready(readable)
            
            # Timers Phase: Handle due timer events
            self._process_timers()

    def _wakeup(self):
        """Wake up the event loop if it is waiting in select."""
        os.write(self.wakeup_write, b'\x00')

    def _calculate_timeout(self):
        """Calculate the time in seconds until the next scheduled timer event."""
        if not self.timer_heap:
            return None
        now = time.time()
        next_time = self.timer_heap[0][0]
        return max(0, next_time - now)

    def _process_timers(self):
        """Process due timer events by calling their callbacks."""
        now = time.time()
        while self.timer_heap and self.timer_heap[0][0] <= now:
            _, callback = heappop(self.timer_heap)
            callback()

    def _process_ready(self, readable):
        """Process ready I/O events by calling their callbacks."""
        for sock in readable:
            if sock != self.wakeup_read and sock in self.readers:
                callback = self.readers[sock]
                callback()

    def add_reader(self, sock, callback):
        """Add a socket and its callback to the readers."""
        self.readers[sock] = callback
        self._wakeup()

    def remove_reader(self, sock):
        """Remove a socket from the readers."""
        if sock in self.readers:
            del self.readers[sock]
            self._wakeup()

    def set_timeout(self, callback, delay):
        """Schedule a one-time callback after a delay."""
        time_to_run = time.time() + delay
        heappush(self.timer_heap, (time_to_run, callback))
        self._wakeup()

    def set_interval(self, callback, interval):
        """Schedule a recurring callback with a specified interval."""
        def wrapper():
            callback()
            self.set_timeout(wrapper, interval)
        self.set_timeout(wrapper, interval)

# Example usage
def on_data_available():
    conn, addr = server_socket.accept()
    data = conn.recv(1024)
    print(f"Received data: {data.decode()}")
    conn.close()

def send_data_periodically():
    client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    client_socket.connect(('localhost', 12345))
    client_socket.sendall(b"Hello, World!")
    client_socket.close()

# Create a server socket
server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
server_socket.bind(('localhost', 12345))
server_socket.listen()

# Create an event loop
loop = EventLoop()
loop.add_reader(server_socket, on_data_available)

# Start sending data periodically using set_interval
loop.set_interval(send_data_periodically, 1)

# Run the event loop (no separate thread, single-threaded operation)
loop.start()
